{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pre-processing mouse single-nuclei RNA-seq data with kallisto and bustools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this tutorial we will process the 10x dataset `1k Brain Nuclei from an E18 Mouse` using kallisto bus and a custom built DNA and intron index for mouse. We will generate two matrices: one for spliced transcripts and one for unspliced transcripts, and sum them to obtain total nuclear transcripts.\n",
    "\n",
    "\n",
    "The 10x dataset `1k Brain Nuclei from an E18 Mouse` (5GB) is available here:\n",
    "\n",
    "https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/nuclei_900\n",
    "\n",
    "To learn how to build a cDNA and intron index see this tutorial:\n",
    "\n",
    "https://www.kallistobus.tools/velocity_index_tutorial.html\n",
    "\n",
    "Important: The mouse cDNA and intron index is about 26GB. Because of this, building it and processing data with it requires significantly more RAM than typical kallisto workflows, and we recoomend using a machine with at least 64GB RAM for this workflow.\n",
    "\n",
    "To save you time, we have made the mouse cDNA and intron index built with the mouse ensembl 86 release. \n",
    "You can download the index and other files used in this tutorial at Caltech Data (19GB zip file) here: \n",
    "\n",
    "==== insert caltech data link to mouse intron index ====\n",
    "```\n",
    " 26GB cDNA_introns.idx\n",
    " 46MB cDNA_introns.t2g.txt\n",
    "3.7MB cDNA_transcripts_to_capture.txt\n",
    " 20MB introns_transcripts_to_capture.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download fastq files and whilelist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2019-08-02 00:07:03--  http://cf.10xgenomics.com/samples/cell-exp/2.1.0/nuclei_900/nuclei_900_fastqs.tar\n",
      "Resolving cf.10xgenomics.com (cf.10xgenomics.com)... 99.84.41.7, 99.84.41.96, 99.84.41.41, ...\n",
      "Connecting to cf.10xgenomics.com (cf.10xgenomics.com)|99.84.41.7|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 5290926080 (4.9G) [application/x-tar]\n",
      "Saving to: ‘nuclei_900_fastqs.tar’\n",
      "\n",
      "100%[====================================>] 5,290,926,080 45.8MB/s   in 2m 40s \n",
      "\n",
      "2019-08-02 00:09:48 (31.5 MB/s) - ‘nuclei_900_fastqs.tar’ saved [5290926080/5290926080]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Download files from 10x genomics and untar\n",
    "!wget http://cf.10xgenomics.com/samples/cell-exp/2.1.0/nuclei_900/nuclei_900_fastqs.tar\n",
    "!tar -xf nuclei_900_fastqs.tar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2019-08-02 00:10:11--  https://github.com/BUStools/getting_started/releases/download/velocity_tutorial/10xv2_whitelist.txt\n",
      "Resolving github.com (github.com)... 192.30.255.112\n",
      "Connecting to github.com (github.com)|192.30.255.112|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/191064839/2f757f00-8d45-11e9-8067-d123e7762f59?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20190802%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190802T071016Z&X-Amz-Expires=300&X-Amz-Signature=b068be06f67ae3ef3da3a48e9f3e55272968c0e203300371e9c1c85acf12c963&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3D10xv2_whitelist.txt&response-content-type=application%2Foctet-stream [following]\n",
      "--2019-08-02 00:10:16--  https://github-production-release-asset-2e65be.s3.amazonaws.com/191064839/2f757f00-8d45-11e9-8067-d123e7762f59?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20190802%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190802T071016Z&X-Amz-Expires=300&X-Amz-Signature=b068be06f67ae3ef3da3a48e9f3e55272968c0e203300371e9c1c85acf12c963&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3D10xv2_whitelist.txt&response-content-type=application%2Foctet-stream\n",
      "Resolving github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)... 52.216.92.19\n",
      "Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)|52.216.92.19|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 12533760 (12M) [application/octet-stream]\n",
      "Saving to: ‘10xv2_whitelist.txt’\n",
      "\n",
      "100%[======================================>] 12,533,760  8.12MB/s   in 1.5s   \n",
      "\n",
      "2019-08-02 00:10:23 (8.12 MB/s) - ‘10xv2_whitelist.txt’ saved [12533760/12533760]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Download 10x Chromium v2 chemistry barcode whitelist 10xv2_whitelist.txt\n",
    "!wget https://github.com/BUStools/getting_started/releases/download/velocity_tutorial/10xv2_whitelist.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 31G\n",
      "-rw-rw-r-- 1 munfred munfred  12M Jun 12 19:14 10xv2_whitelist.txt\n",
      "-rw-rw-r-- 1 munfred munfred  26G Aug  1 16:17 cDNA_introns.idx\n",
      "-rw-rw-r-- 1 munfred munfred  46M Aug  1 16:17 cDNA_introns.t2g.txt\n",
      "-rw-rw-r-- 1 munfred munfred 3.7M Aug  1 16:17 cDNA_transcripts_to_capture.txt\n",
      "-rw-rw-r-- 1 munfred munfred  20M Aug  1 16:17 introns_transcripts_to_capture.txt\n",
      "-rw-rw-r-- 1 munfred munfred 3.6K Aug  2 00:09 kallisto_bus_mouse_nuclei_tutorial.ipynb\n",
      "drwxr-xr-x 2 munfred munfred 4.0K Aug 23  2017 nuclei_900_fastqs\n",
      "-rw-rw-r-- 1 munfred munfred 5.0G Nov  8  2017 nuclei_900_fastqs.tar\n"
     ]
    }
   ],
   "source": [
    "# make sure all files were downloaded\n",
    "!ls -lh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 5.0G\n",
      "-rw-r--r-- 1 munfred munfred 225M Aug 23  2017 nuclei_900_S1_L001_I1_001.fastq.gz\n",
      "-rw-r--r-- 1 munfred munfred 599M Aug 23  2017 nuclei_900_S1_L001_R1_001.fastq.gz\n",
      "-rw-r--r-- 1 munfred munfred 1.7G Aug 23  2017 nuclei_900_S1_L001_R2_001.fastq.gz\n",
      "-rw-r--r-- 1 munfred munfred 226M Aug 23  2017 nuclei_900_S1_L002_I1_001.fastq.gz\n",
      "-rw-r--r-- 1 munfred munfred 602M Aug 23  2017 nuclei_900_S1_L002_R1_001.fastq.gz\n",
      "-rw-r--r-- 1 munfred munfred 1.7G Aug 23  2017 nuclei_900_S1_L002_R2_001.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "!ls -lh ./nuclei_900_fastqs/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m.\u001b[00m\n",
      "├── 10xv2_whitelist.txt\n",
      "├── cDNA_introns.idx\n",
      "├── cDNA_introns.t2g.txt\n",
      "├── cDNA_transcripts_to_capture.txt\n",
      "├── introns_transcripts_to_capture.txt\n",
      "├── kallisto_bus_mouse_nuclei_tutorial.ipynb\n",
      "├── \u001b[01;34mnuclei_900_fastqs\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L001_I1_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L001_R1_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L001_R2_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L002_I1_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L002_R1_001.fastq.gz\u001b[00m\n",
      "│   └── \u001b[01;31mnuclei_900_S1_L002_R2_001.fastq.gz\u001b[00m\n",
      "└── \u001b[01;31mnuclei_900_fastqs.tar\u001b[00m\n",
      "\n",
      "1 directory, 13 files\n"
     ]
    }
   ],
   "source": [
    "!tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run kallisto  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "kallisto 0.46.0\n",
      "Generates BUS files for single-cell sequencing\n",
      "\n",
      "Usage: kallisto bus [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              pseudoalignment\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "-x, --technology=STRING       Single-cell technology used \n",
      "\n",
      "Optional arguments:\n",
      "-l, --list                    List all single-cell technologies supported\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n"
     ]
    }
   ],
   "source": [
    "!kallisto bus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 818,724\n",
      "[index] number of k-mers: 1,105,269,838\n",
      "[index] number of equivalence classes: 5,740,477\n",
      "[quant] will process sample 1: nuclei_900_fastqs/nuclei_900_S1_L001_R1_001.fastq.gz\n",
      "                               nuclei_900_fastqs/nuclei_900_S1_L001_R2_001.fastq.gz\n",
      "[quant] will process sample 2: nuclei_900_fastqs/nuclei_900_S1_L002_R1_001.fastq.gz\n",
      "                               nuclei_900_fastqs/nuclei_900_S1_L002_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 48,909,315 reads, 44,705,830 reads pseudoaligned\n"
     ]
    }
   ],
   "source": [
    "!kallisto bus -i cDNA_introns.idx -o bus_output -x 10xv2 -t 4 nuclei_900_fastqs/nuclei_900_S1_L001_R1_001.fastq.gz nuclei_900_fastqs/nuclei_900_S1_L001_R2_001.fastq.gz  nuclei_900_fastqs/nuclei_900_S1_L002_R1_001.fastq.gz nuclei_900_fastqs/nuclei_900_S1_L002_R2_001.fastq.gz "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m.\u001b[00m\n",
      "├── 10xv2_whitelist.txt\n",
      "├── \u001b[01;34mbus_output\u001b[00m\n",
      "│   ├── matrix.ec\n",
      "│   ├── output.bus\n",
      "│   ├── run_info.json\n",
      "│   └── transcripts.txt\n",
      "├── cDNA_introns.idx\n",
      "├── cDNA_introns.t2g.txt\n",
      "├── cDNA_transcripts_to_capture.txt\n",
      "├── introns_transcripts_to_capture.txt\n",
      "├── kallisto_bus_mouse_nuclei_tutorial.ipynb\n",
      "├── \u001b[01;34mnuclei_900_fastqs\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L001_I1_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L001_R1_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L001_R2_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L002_I1_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L002_R1_001.fastq.gz\u001b[00m\n",
      "│   └── \u001b[01;31mnuclei_900_S1_L002_R2_001.fastq.gz\u001b[00m\n",
      "└── \u001b[01;31mnuclei_900_fastqs.tar\u001b[00m\n",
      "\n",
      "2 directories, 17 files\n"
     ]
    }
   ],
   "source": [
    "# check output files\n",
    "!tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run bustools\n",
    "Correct, sort, capture, and count the spliced and unspliced matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bustools 0.39.3\n",
      "\n",
      "Usage: bustools <CMD> [arguments] ..\n",
      "\n",
      "Where <CMD> can be one of: \n",
      "\n",
      "capture         Capture records from a BUS file\n",
      "correct         Error correct a BUS file\n",
      "count           Generate count matrices from a BUS file\n",
      "inspect         Produce a report summarizing a BUS file\n",
      "linker          Remove section of barcodes in BUS files\n",
      "project         Project a BUS file to gene sets\n",
      "sort            Sort a BUS file by barcodes and UMIs\n",
      "text            Convert a binary BUS file to a tab-delimited text file\n",
      "whitelist       Generate a whitelist from a BUS file\n",
      "\n",
      "Running bustools <CMD> without arguments prints usage information for <CMD>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!bustools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p bus_output/cDNA_capture/ bus_output/intron_capture/ bus_output/spliced/ bus_output/unspliced/ bus_output/tmp/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 737280 barcodes in the whitelist\n",
      "Number of hamming dist 1 barcodes = 20550336\n",
      "Processed 44705830 bus records\n",
      "In whitelist = 43426576\n",
      "Corrected = 305038\n",
      "Uncorrected = 974216\n"
     ]
    }
   ],
   "source": [
    "!bustools393 correct -w 10xv2_whitelist.txt -o bus_output/output.correct.bus bus_output/output.bus "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in 43731614 BUS records\n"
     ]
    }
   ],
   "source": [
    "!bustools393 sort -o bus_output/output.correct.sort.bus -t 4 bus_output/output.correct.bus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 5.1G\n",
      "drwxrwxr-x 7 munfred munfred 4.0K Aug  2 00:52 .\n",
      "drwxrwxr-x 5 munfred munfred 4.0K Aug  2 01:13 ..\n",
      "drwxrwxr-x 2 munfred munfred 4.0K Aug  2 00:31 cDNA_capture\n",
      "drwxrwxr-x 2 munfred munfred 4.0K Aug  2 00:31 introns_capture\n",
      "-rw-rw-r-- 1 munfred munfred 1.7G Aug  2 00:30 matrix.ec\n",
      "-rw-rw-r-- 1 munfred munfred 1.4G Aug  2 00:30 output.bus\n",
      "-rw-rw-r-- 1 munfred munfred 1.4G Aug  2 00:51 output.correct.bus\n",
      "-rw-rw-r-- 1 munfred munfred 817M Aug  2 00:52 output.correct.sort.bus\n",
      "-rw-rw-r-- 1 munfred munfred  546 Aug  2 00:30 run_info.json\n",
      "drwxrwxr-x 2 munfred munfred 4.0K Aug  2 00:31 spliced\n",
      "drwxrwxr-x 2 munfred munfred 4.0K Aug  2 00:31 tmp\n",
      "-rw-rw-r-- 1 munfred munfred  24M Aug  2 00:30 transcripts.txt\n",
      "drwxrwxr-x 2 munfred munfred 4.0K Aug  2 00:31 unspliced\n"
     ]
    }
   ],
   "source": [
    "!ls -lah bus_output/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m.\u001b[00m\n",
      "├── 10xv2_whitelist.txt\n",
      "├── \u001b[01;34mbus_output\u001b[00m\n",
      "│   ├── \u001b[01;34mcDNA_capture\u001b[00m\n",
      "│   ├── \u001b[01;34mintrons_capture\u001b[00m\n",
      "│   ├── matrix.ec\n",
      "│   ├── output.bus\n",
      "│   ├── output.correct.bus\n",
      "│   ├── output.correct.sort.bus\n",
      "│   ├── run_info.json\n",
      "│   ├── \u001b[01;34mspliced\u001b[00m\n",
      "│   ├── \u001b[01;34mtmp\u001b[00m\n",
      "│   ├── transcripts.txt\n",
      "│   └── \u001b[01;34munspliced\u001b[00m\n",
      "├── cDNA_introns.idx\n",
      "├── cDNA_introns_t2g.txt\n",
      "├── cDNA_transcripts_to_capture.txt\n",
      "├── introns_transcripts_to_capture.txt\n",
      "├── kallisto_bus_mouse_nuclei_tutorial.ipynb\n",
      "├── \u001b[01;34mnuclei_900_fastqs\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L001_I1_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L001_R1_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L001_R2_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L002_I1_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L002_R1_001.fastq.gz\u001b[00m\n",
      "│   └── \u001b[01;31mnuclei_900_S1_L002_R2_001.fastq.gz\u001b[00m\n",
      "└── \u001b[01;31mnuclei_900_fastqs.tar\u001b[00m\n",
      "\n",
      "7 directories, 19 files\n"
     ]
    }
   ],
   "source": [
    "!tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsing transcripts .. done\n",
      "Parsing ECs .. done\n",
      "Parsing capture list .. done\n",
      "Read in 26766408 BUS records, wrote 22230247 BUS records\n"
     ]
    }
   ],
   "source": [
    "!bustools393 capture -s -o bus_output/cDNA_capture/cDNA_capture.bus -c cDNA_transcripts_to_capture.txt -e bus_output/matrix.ec -t bus_output/transcripts.txt bus_output/output.correct.sort.bus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsing transcripts .. done\n",
      "Parsing ECs .. done\n",
      "Parsing capture list .. done\n",
      "Read in 26766408 BUS records, wrote 11835626 BUS records\n"
     ]
    }
   ],
   "source": [
    "!bustools393 capture -s -o bus_output/intron_capture/intron_capture.bus -c introns_transcripts_to_capture.txt -e bus_output/matrix.ec -t bus_output/transcripts.txt bus_output/output.correct.sort.bus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 679M\n",
      "-rw-rw-r-- 1 munfred munfred 679M Aug  2 11:02 cDNA_capture.bus\n",
      "total 362M\n",
      "-rw-rw-r-- 1 munfred munfred 362M Aug  2 11:13 intron_capture.bus\n"
     ]
    }
   ],
   "source": [
    "!ls -lh ./bus_output/cDNA_capture\n",
    "!ls -lh ./bus_output/intron_capture"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m.\u001b[00m\n",
      "├── 10xv2_whitelist.txt\n",
      "├── \u001b[01;34mbus_output\u001b[00m\n",
      "│   ├── \u001b[01;34mcDNA_capture\u001b[00m\n",
      "│   │   └── cDNA_capture.bus\n",
      "│   ├── \u001b[01;34mintron_capture\u001b[00m\n",
      "│   │   └── intron_capture.bus\n",
      "│   ├── matrix.ec\n",
      "│   ├── output.bus\n",
      "│   ├── output.correct.bus\n",
      "│   ├── output.correct.sort.bus\n",
      "│   ├── run_info.json\n",
      "│   ├── \u001b[01;34mspliced\u001b[00m\n",
      "│   ├── \u001b[01;34mtmp\u001b[00m\n",
      "│   ├── transcripts.txt\n",
      "│   └── \u001b[01;34munspliced\u001b[00m\n",
      "│       └── \u001b[01;34mu\u001b[00m\n",
      "├── cDNA_introns.idx\n",
      "├── cDNA_introns_t2g.txt\n",
      "├── cDNA_transcripts_to_capture.txt\n",
      "├── introns_transcripts_to_capture.txt\n",
      "├── kallisto_bus_mouse_nuclei_tutorial.ipynb\n",
      "├── \u001b[01;34mnuclei_900_fastqs\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L001_I1_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L001_R1_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L001_R2_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L002_I1_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L002_R1_001.fastq.gz\u001b[00m\n",
      "│   └── \u001b[01;31mnuclei_900_S1_L002_R2_001.fastq.gz\u001b[00m\n",
      "└── \u001b[01;31mnuclei_900_fastqs.tar\u001b[00m\n",
      "\n",
      "8 directories, 21 files\n"
     ]
    }
   ],
   "source": [
    "!tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bustools393 count -o bus_output/unspliced/unspliced -g cDNA_introns_t2g.txt -e bus_output/matrix.ec -t bus_output/transcripts.txt --genecounts bus_output/cDNA_capture/cDNA_capture.bus "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bustools393 count -o bus_output/spliced/spliced -g cDNA_introns_t2g.txt -e bus_output/matrix.ec -t bus_output/transcripts.txt --genecounts bus_output/intron_capture/intron_capture.bus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m.\u001b[00m\n",
      "├── 10xv2_whitelist.txt\n",
      "├── \u001b[01;34mbus_output\u001b[00m\n",
      "│   ├── \u001b[01;34mcDNA_capture\u001b[00m\n",
      "│   │   └── cDNA_capture.bus\n",
      "│   ├── \u001b[01;34mintron_capture\u001b[00m\n",
      "│   │   └── intron_capture.bus\n",
      "│   ├── matrix.ec\n",
      "│   ├── output.bus\n",
      "│   ├── output.correct.bus\n",
      "│   ├── output.correct.sort.bus\n",
      "│   ├── run_info.json\n",
      "│   ├── \u001b[01;34mspliced\u001b[00m\n",
      "│   │   ├── \u001b[01;34mspliced\u001b[00m\n",
      "│   │   ├── spliced.barcodes.txt\n",
      "│   │   ├── spliced.genes.txt\n",
      "│   │   └── spliced.mtx\n",
      "│   ├── \u001b[01;34mtmp\u001b[00m\n",
      "│   ├── transcripts.txt\n",
      "│   └── \u001b[01;34munspliced\u001b[00m\n",
      "│       ├── \u001b[01;34munspliced\u001b[00m\n",
      "│       ├── unspliced.barcodes.txt\n",
      "│       ├── unspliced.genes.txt\n",
      "│       └── unspliced.mtx\n",
      "├── cDNA_introns.idx\n",
      "├── cDNA_introns_t2g.txt\n",
      "├── cDNA_transcripts_to_capture.txt\n",
      "├── introns_transcripts_to_capture.txt\n",
      "├── kallisto_bus_mouse_nuclei_tutorial.ipynb\n",
      "├── \u001b[01;34mnuclei_900_fastqs\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L001_I1_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L001_R1_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L001_R2_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L002_I1_001.fastq.gz\u001b[00m\n",
      "│   ├── \u001b[01;31mnuclei_900_S1_L002_R1_001.fastq.gz\u001b[00m\n",
      "│   └── \u001b[01;31mnuclei_900_S1_L002_R2_001.fastq.gz\u001b[00m\n",
      "└── \u001b[01;31mnuclei_900_fastqs.tar\u001b[00m\n",
      "\n",
      "9 directories, 27 files\n"
     ]
    }
   ],
   "source": [
    "!tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load spliced and unspliced matrices in Python and merge them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from anndata import AnnData\n",
    "import anndata\n",
    "from scipy import sparse\n",
    "import scipy\n",
    "import anndata\n",
    "import scipy.io\n",
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded unspliced count matrix.\n",
      "AnnData object with n_obs × n_vars = 223419 × 54838 \n"
     ]
    }
   ],
   "source": [
    "## load unspliced data on anndata as sparse crs matrix\n",
    "unspliced = anndata.AnnData(scipy.io.mmread('./bus_output/unspliced/unspliced.mtx').tocsr())\n",
    "unspliced.obs= pd.read_csv('./bus_output/unspliced/unspliced.barcodes.txt', index_col = 0, header = None, names = ['barcode'])\n",
    "unspliced.var = pd.read_csv('./bus_output/unspliced/unspliced.genes.txt', header = None, index_col = 0, names =['ensembl_id'], sep = '\\t')\n",
    "print('Loaded unspliced count matrix.')\n",
    "print(unspliced)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded spliced count matrix\n",
      "AnnData object with n_obs × n_vars = 170857 × 54838 \n"
     ]
    }
   ],
   "source": [
    "## load unspliced data on anndata as sparse crs matrix\n",
    "spliced = anndata.AnnData(scipy.io.mmread('./bus_output/spliced/spliced.mtx').tocsr())\n",
    "spliced.obs= pd.read_csv('./bus_output/spliced/spliced.barcodes.txt', index_col = 0, header = None, names = ['barcode'])\n",
    "spliced.var = pd.read_csv('./bus_output/spliced/spliced.genes.txt', header = None, index_col = 0, names =['ensembl_id'], sep = '\\t')\n",
    "print('Loaded spliced count matrix')\n",
    "print(spliced)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 223419 × 54838 \n"
     ]
    }
   ],
   "source": [
    "print(unspliced)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAACGAG</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAAGTGG</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGCCCA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGGTTT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAATTCCC</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [AAACCTGAGAAACGAG, AAACCTGAGAAAGTGG, AAACCTGAGAAGCCCA, AAACCTGAGAAGGTTT, AAACCTGAGAATTCCC]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spliced.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102693.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000064842.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000051951.5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102851.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000103377.1</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [ENSMUSG00000102693.1, ENSMUSG00000064842.1, ENSMUSG00000051951.5, ENSMUSG00000102851.1, ENSMUSG00000103377.1]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spliced.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAACGCC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAAGTGG</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGCCCA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAATTCCC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGACACGAC</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [AAACCTGAGAAACGCC, AAACCTGAGAAAGTGG, AAACCTGAGAAGCCCA, AAACCTGAGAATTCCC, AAACCTGAGACACGAC]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unspliced.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102693.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000064842.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000051951.5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102851.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000103377.1</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [ENSMUSG00000102693.1, ENSMUSG00000064842.1, ENSMUSG00000051951.5, ENSMUSG00000102851.1, ENSMUSG00000103377.1]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unspliced.var.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sum spliced + unspliced counts\n",
    "Now that we have spliced and unspliced matrices we can sum the counts of genes for barcodes common to both matrices\n",
    "We take the intersection of both matrices because presumably cells without a single count on either have very low counts anyway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "idx = spliced.obs.index.intersection(unspliced.obs.index)\n",
    "spliced_intersection = spliced[idx]\n",
    "spliced_intersection = unspliced[idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<139214x54838 sparse matrix of type '<class 'numpy.float32'>'\n",
       "\twith 6464835 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spliced_intersection.X + unspliced_intersection.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 139214 × 54838 "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spliced_plus_unspliced = spliced_intersection.copy()\n",
    "spliced_plus_unspliced.X = spliced_intersection.X + unspliced_intersection.X\n",
    "spliced_plus_unspliced"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
